### PR TITLE
feat: typescript support [CU-257cp4g]

### PR DESCRIPTION
**Pull request by ltsNotMike**
## Summary

What does this PR do/solve? 

> Adds the possibility to write the plugin in typescript

## Test Plan

How are you testing the work you're submitting?

- Running publish pipeline locally and check if the package contains everything
- Link locally published package to strapi project and check if the plugin works
- Run development build, link the build folder to strapi project, and check if the plugin works
- In all tests, I've checked if .ts files are being compiled

### CODE REVIEW

**Comment by cyp3rius** [comment]
Task linked: [CU-257cp4g TypeScript support](https://app.clickup.com/t/257cp4g)

File: server/utils/functions.ts [hunk 1]
```diff
@@ -0,0 +1,298 @@
...
+      [curr]: parsedValue,
+    };
+  }, {});
+
+
+
+export const templateNameFactory = async (
+  items: Array<NavigationItemEntity<ContentTypeEntity[] | ContentTypeEntity>> = [],
+  strapi: IStrapi,
+  contentTypes: Array<StrapiContentType<any>> = [],
```
Comment (CodeVoyager) [inline server/utils/functions.ts:45]: If we really need `any` please at least let's mark those places with `ToBeFixed`

File: server/services/index.ts [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+
+import commonService from './common';
+import adminService from './admin';
+import clientService from './client';
+
+const navigationService = {
+  common: commonService,
+  admin: adminService,
+  client: clientService,
+};
```
Comment (CodeVoyager) [inline server/services/index.ts:11]:
```suggestion
import common from './common';
import admin from './admin';
import client from './client';

const navigationService = {
  common,
  admin,
  client,
};
```

🤷

File: server/services/common.ts [hunk 3]
```diff
@@ -0,0 +1,394 @@
...
+		const commonService = getPluginService<ICommonService>('common');
+		const { toCreate, toRemove, toUpdate } = items
+			.reduce((acc: NavigationActionsPerItem, _) => {
+				const branchName: keyof NavigationActionsPerItem | void = commonService.getBranchName(_);
+				if (branchName) {
+					return { ...acc, [branchName]: [...acc[branchName], _] };
+				}
+				return acc;
+			},
+				{ toRemove: [], toCreate: [], toUpdate: [] },
```
Comment (CodeVoyager) [inline server/services/common.ts:23]: Yet another mixup of tabs and spaces.

File: server/services/client.ts [hunk 4]
```diff
@@ -0,0 +1,313 @@
...
+		const clientService = getPluginService<IClientService>('client');
+
+		const findById = !isNaN(toNumber(idOrSlug)) || validate(idOrSlug as string);
+		const criteria = findById ? { id: idOrSlug } : { slug: idOrSlug };
+		const itemCriteria = menuOnly ? { menuAttached: true } : {};
+		return await clientService.renderType(type, criteria, itemCriteria, null, rootPath);
+	},
+
+	async renderChildren(
+      idOrSlug: Id | string,
```
Comment (CodeVoyager) [inline server/services/client.ts:26]: Since `Id` is `string | number` I'm not sure if we need extra `| string`

File: server/controllers/index.ts [hunk 5]
```diff
@@ -0,0 +1,11 @@
+import { NavigationController } from '../../types';
+
+import adminControllers from './admin';
+import clientControllers from './client';
+
+const controllers: NavigationController = {
+  admin: adminControllers,
+  client: clientControllers,
+}
```
Comment (CodeVoyager) [inline server/controllers/index.ts:9]:
```suggestion
import admin from './admin';
import client from './client';

const controllers: NavigationController = {
  admin,
  client,
}
```

🤷

**Review by cyp3rius — APPROVED** [review]

**Review by CodeVoyager — APPROVED** [review]

**Comment by cyp3rius** [comment]
Rebase done. Worth to test it @ltsNotMike.

**Pull request merged**